lib/membrane_rtsp_plugin/tcp_decapsulator.ex

```diff
@@ -0,0 +1,121 @@
+
+  alias Membrane.{Buffer, RemoteStream, RTP, RTSP}
+
+  def_options rtsp_session: [
+                spec: pid() | nil,
+                default: nil,
+                description: """
+                PID of a RTSP Session (returned from Membrane.RTSP.start or Membrane.RTSP.start_link)
+                that received RTSP responses will be forwarded to. If nil the responses will be
+                discarded.
+                """
+              ]
+
+  def_input_pad :input, accepted_format: %RemoteStream{type: :bytestream}
+
+  def_output_pad :output, accepted_format: %RemoteStream{type: :packetized, content_format: RTP}
+
+  @impl true
+  def handle_init(_ctx, opts) do
+    state =
+      Map.from_struct(opts)
+      |> Map.merge(%{
+        unprocessed_data: <<>>
+      })
+
+    {[], state}
+  end
+
+  @impl true
+  def handle_playing(_ctx, state) do
+    stream_format = %RemoteStream{type: :packetized, content_format: RTP}
+    {[stream_format: {:output, stream_format}], state}
+  end
+
+  @impl true
+  def handle_stream_format(:input, _stream_format, _ctx, state) do
+    {[], state}
+  end
+
+  @impl true
+  def handle_buffer(:input, %Buffer{payload: payload, metadata: metadata}, _ctx, state) do
+    packets_binary = state.unprocessed_data <> payload
+
+    {unprocessed_data, complete_packets_binaries} =
+      get_complete_packets(packets_binary, state.rtsp_session)
+
+    packets_buffers =
+      Enum.map(complete_packets_binaries, &%Buffer{payload: &1, metadata: metadata})
+
+    {[buffer: {:output, packets_buffers}], %{state | unprocessed_data: unprocessed_data}}
+  end
+
+  @spec get_complete_packets(binary(), pid() | nil, [binary()]) ::
+          {unprocessed_data :: binary(), complete_packets :: [binary()]}
+  defp get_complete_packets(packets_binary, rtsp_session, complete_packets \\ [])
+
+  defp get_complete_packets(packets_binary, _rtsp_session, complete_packets)
+       when byte_size(packets_binary) <= 4 do
+    {packets_binary, Enum.reverse(complete_packets)}
+  end
+
+  defp get_complete_packets(
+         <<"$", _channel_id, payload_length::size(16), rest::binary>> = packets_binary,
+         rtsp_session,
+         complete_packets
+       ) do
+    case rest do
+      <<complete_packet_binary::binary-size(payload_length), rest::binary>> ->
+        complete_packets = [complete_packet_binary | complete_packets]
+
+        get_complete_packets(rest, rtsp_session, complete_packets)
+
+      _incomplete_packet_binary ->
+        {packets_binary, Enum.reverse(complete_packets)}
+    end
+  end
+
+  defp get_complete_packets(
+         <<"RTSP", _rest::binary>> = rtsp_message_start,
+         rtsp_session,
+         complete_packets_binaries
+       ) do
+    case RTSP.Response.verify_content_length(rtsp_message_start) do
+      {:ok, _expected_length, _actual_length} ->
+        if rtsp_session != nil do
+          {:ok, %RTSP.Response{status: 200}} =
+            RTSP.handle_response(rtsp_session, rtsp_message_start)
```

Comment:
Tried this on the axis camera and I got this
```elixir
17:55:13.490 [error] GenServer #PID<0.1195.0> terminating
** (MatchError) no match of right hand side value: {:ok, %Membrane.RTSP.Response{status: 401, version: "1.0", headers: [{"CSeq", "19"}, {"WWW-Authenticate", "Digest realm=\"AXIS_WS_B8A44FA90510\", nonce=\"000218d8Y193731bd77ec9811117e74cba1d88a456f9ac2\", stale=TRUE"}, {"Server", "GStreamer RTSP server"}, {"Session", "oN+e30GvGL43gFFN;timeout=60"}, {"Date", "Tue, 11 Jun 2024 16:55:13 GMT"}], body: ""}}
    (membrane_rtsp_plugin 0.2.0) lib/membrane_rtsp_plugin/tcp_decapsulator.ex:98: Membrane.RTSP.TCP.Decapsulator.get_complete_packets/3
    (membrane_rtsp_plugin 0.2.0) lib/membrane_rtsp_plugin/tcp_decapsulator.ex:57: Membrane.RTSP.TCP.Decapsulator.handle_buffer/4
    (membrane_core 1.0.1) lib/membrane/core/callback_handler.ex:139: Membrane.Core.CallbackHandler.exec_callback/4
    (membrane_core 1.0.1) lib/membrane/core/callback_handler.ex:69: Membrane.Core.CallbackHandler.exec_and_handle_callback/5
    (elixir 1.15.4) lib/enum.ex:2510: Enum."-reduce/3-lists^foldl/2-0-"/3
    (membrane_core 1.0.1) lib/membrane/core/element.ex:232: Membrane.Core.Element.handle_info/2
    (stdlib 5.2.1) gen_server.erl:1095: :gen_server.try_handle_info/3
    (stdlib 5.2.1) gen_server.erl:1183: :gen_server.handle_msg/6
Last message: {Membrane.Core.Message, :buffer, [%Membrane.Buffer{payload: <<36, 254, 12, 83, 195, 145, 54, 218, 122, 208, 224, 19, 67, 59, 226, 25, 254, 207, 195, 5, 136, 119, 243, 29, 40, 117, 41, 248, 170, 7, 2, 21, 160, 34, 175, 99, 247, 5, 89, 85, 10, 128, 128, 179, 185, ...>>, pts: nil, dts: nil, metadata: %{tcp_source_address: {xx, xx, xx, xx}, tcp_source_port: xx, arrival_ts: 1718124913485083938}}], [for_pad: :input]}
...
```
So the session crash before the rtsp session will have the chance to refresh the credentials.

Reply:
I think we ultimately should keep the credentials refreshed, I added simple temporary logic that would rerequest GET_PARAMETER in case an interleaved 401 response is captured, could you try it now?

Reply:
Yes, I think this fixes the issue